2-transform-data/bq_ingestion/extract_msgs.py

```diff
@@ -26,351 +26,428 @@
-        break
-  else:
-    body = msg.get_payload(decode=True)
-  mparts = msg.items()
-  # currently, writing both the decoded string and the base64-ified bytestring for the body,
-  # to bigquery. (Did this b/c I wasn't confident in how the decoding was happening in all cases.
-  # However, I think it may be essentially redundant.)
-  mparts.append(('Body', try_decode(body)))
-  b64_bstring = base64.b64encode(body)
-  b64_string = (str(b64_bstring)[2:])[:-1]  # uhhh... there has got to be a better way to convert
-  # the base64 bytestring to a a string, needed for the json bq ingestion (but I couldn't figure
-  # it out)
-  mparts.append(('body_bytes', b64_string))
-  return mparts
-
-
-# TODO: this essentially works... but what's the best way to deal with all these different formats?
-# (update: after discn on internal python chat channel, seems this may be the best approach...
-# All the different formats are probably due to ancient mail client variants. It's the older
-# messages that have issues.)
+            return blob.decode(codec)
+        except (UnicodeDecodeError, LookupError, AttributeError) as e:
+            err = e
+    print('Cannot decode string: {} based on error: {}\n'.format(blob, err))
+    raise err
+
+def decompress_line_by_line(blob, fpath, split_regex_value):
+    # Include timestamp in local file name to avoid archive name clashes in case we're processing multiple buckets concurrently.
+    message_lines, messages_list_result = [], []
+
+    # Pull down file locally to loop over
+    temp_file = '/tmp/{}_{}'.format(int(time.time()), fpath)
+    blob.download_to_filename(temp_file)
+
+    try:
+        with gzip.open(temp_file, 'rb') as encoded_file:
+            for line in encoded_file:
+                # Decode the line for re.search() call
+                try:
+                    decode_line = decode_messsage(line)
+                # TODO: Unlikely to occur but catching if it does
+                except UnicodeDecodeError as e:
+                    print("{} error decoding line {}".format(e, line))
+                    break
+                split_point = re.search(split_regex_value, decode_line)
+                if split_point:
+                    # Combine message contents prior to From
+                    messages_list_result.append(b''.join(message_lines))
+                    # Place full msg into list
+                    message_lines = [line]
+                else:
+                    message_lines.append(line)
+            # Catches last message
+            if message_lines:
+                messages_list_result.append(b''.join(message_lines))
+    except e:
+        print(e)
+        print('{} not successfully gunzipped'.format(fpath))
+    # Delete temp file
+    finally:
+        if os.path.exists(temp_file):
+            os.remove(temp_file)
+
+    return messages_list_result
+
+# TODO: this will break if the 'filename' includes a path. Check for/create parent dir first.
```

Comment:
forgot to mention -- this comment is related to creating the local temp file, so it can be moved to line 68 above in `decompress_line_by_line`